internal/resources/azure/mssql_managed_instance.go

```diff
@@ -0,0 +1,202 @@
+
+	"github.com/infracost/infracost/internal/resources"
+	"github.com/infracost/infracost/internal/schema"
+	"github.com/shopspring/decimal"
+)
+
+const (
+	mssqlMIServiceName   = "MSSQL Managed Instance"
+	mssqlMIProductFamily = "Databases"
+)
+
+// MSSQLManagedInstance struct represents an azure Sql Managed Instance.
+//
+// # MSSQLManagedInstance currently only Gen5 database instance
+//
+// More resource information here: https://azure.microsoft.com/en-gb/products/azure-sql/managed-instance/
+// Pricing information here: https://azure.microsoft.com/en-gb/pricing/details/azure-sql-managed-instance/single/
+type MSSQLManagedInstance struct {
+	Address            string
+	Region             string
+	SKU                string
+	LicenseType        string
+	Cores              int64
+	StorageSizeInGb    int64
+	StorageAccountType string
+	// LongTermRetentionStorageGB defines a usage param that allows users to define how many gb of cold storage the database uses.
+	// This is storage that can be kept for up to 10 years.
+	LongTermRetentionStorageGB *int64 `infracost_usage:"long_term_retention_storage_gb"`
+	BackupStorageGB            *int64 `infracost_usage:"backup_storage_gb"`
+}
+
+// PopulateUsage parses the u schema.UsageData into the MSSQLManagedInstance.
+// It uses the `infracost_usage` struct tags to populate data into the MSSQLManagedInstance.
+func (r *MSSQLManagedInstance) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+// BuildResource builds a schema.Resource from a valid MSSQLManagedInstance struct.
+// This method is called after the resource is initialised by an IaC provider.
+// See providers folder for more information.
+func (r *MSSQLManagedInstance) BuildResource() *schema.Resource {
+	costComponents := r.costComponents()
+
+	return &schema.Resource{
+		Name: r.Address,
+		UsageSchema: []*schema.UsageItem{
+			{Key: "backup_storage_gb", DefaultValue: 0, ValueType: schema.Int64},
+			{Key: "long_term_retention_storage_gb", DefaultValue: 0, ValueType: schema.Int64},
+		},
+		CostComponents: costComponents,
+	}
+}
+
+func (r *MSSQLManagedInstance) costComponents() []*schema.CostComponent {
+	costComponents := []*schema.CostComponent{
+		{
+			Name:           fmt.Sprintf("Compute (%s %d Cores)", strings.ToTitle(r.SKU), r.Cores),
+			Unit:           "hours",
+			UnitMultiplier: decimal.NewFromInt(1),
+			HourlyQuantity: decimalPtr(decimal.NewFromInt(1)),
+			ProductFilter: &schema.ProductFilter{
+				VendorName:    strPtr(vendorName),
+				Region:        strPtr(r.Region),
+				Service:       strPtr(mssqlMIServiceName),
+				ProductFamily: strPtr(mssqlMIProductFamily),
+				AttributeFilters: ([]*schema.AttributeFilter{
+					{Key: "productName", Value: r.productDescription()},
+					{Key: "skuName", Value: r.meteredName()},
+				}),
+			},
+			PriceFilter: priceFilterConsumption,
+		},
+	}
+
+	costComponents = append(costComponents, r.mssqlMIStorageCostComponent(), r.mssqlMIBackupCostComponent())
```

Comment:
We should probably only add the storage cost component if it is above the default
```suggestion
	if r.StorageSizeInGb - 32 > 0 {
		costComponents = append(costComponents, r.mssqlMIStorageCostComponent(), r.mssqlMIBackupCostComponent())
	}
```